src/main/scala/fi/oph/koski/massaluovutus/valintalaskenta/ValintalaskentaQuery.scala

```diff
@@ -50,22 +55,60 @@ case class ValintalaskentaQuery(
       suoritustyypit = suoritustyypit.orElse(Some(ValintalaskentaQuery.defaultSuoritustyypit)),
     ))
 
-  private def getOpiskeluoikeus(application: KoskiApplication, oppijaOid: String): Seq[ValintalaskentaOpiskeluoikeus] = {
+  private def getOpiskeluoikeus(application: KoskiApplication, oppijaOid: String)(implicit user: KoskiSpecificSession): Seq[ValintalaskentaOpiskeluoikeus] =
+    getOpiskeluoikeuksienVersiot(application.masterDatabase.db, oppijaOid).flatMap { oo =>
+      if (oo.rajapäivänVersio == oo.uusinVersio) {
+        getUusinOpiskeluoikeus(application, oo.id)
+      } else if (oo.rajapäivänVersio > 0) {
+        getOpiskeluoikeusHistoriasta(application, oo.oid, oo.rajapäivänVersio)
+      } else {
+        None
+      }
+    }
+
+  private def getUusinOpiskeluoikeus(application: KoskiApplication, id: Int): Option[ValintalaskentaOpiskeluoikeus] =
     QueryMethods.runDbSync(application.masterDatabase.db, sql"""
+          SELECT *
+          FROM opiskeluoikeus
+          WHERE ID=$id
+        """.as[KoskiOpiskeluoikeusRow])
+      .headOption
+      .flatMap(toResponse(application))
+
+  private def getOpiskeluoikeusHistoriasta(application: KoskiApplication, oid: String, versio: Int)(implicit user: KoskiSpecificSession): Option[ValintalaskentaOpiskeluoikeus] = {
+    (application.historyRepository.findVersion(oid, versio) match {
+      case Right(row) =>
+        Some(row)
+      case Left(error) =>
+        logger.error(s"Opiskeluoikeushistoriasta haku epäonnistui: $error")
+        None
+    }).map(ValintalaskentaOpiskeluoikeus.apply)
+  }
+
+  private def getOpiskeluoikeuksienVersiot(db: DB, oppijaOid: String): Seq[OpiskeluoikeudenVersiot] =
+    QueryMethods.runDbSync(db, sql"""
       WITH kaikki_oidit AS (
         SELECT oid
         FROM henkilo
         WHERE oid = $oppijaOid OR master_oid = $oppijaOid
       )
-      SELECT *
+      SELECT
+        opiskeluoikeus.id,
+        opiskeluoikeus.oid,
+        opiskeluoikeus.versionumero AS uusin_versio,
+        max(opiskeluoikeushistoria.versionumero) AS rajapaivan_versio
       FROM opiskeluoikeus
-	    LEFT JOIN kaikki_oidit ON kaikki_oidit.oid = opiskeluoikeus.oppija_oid
+      LEFT JOIN kaikki_oidit ON kaikki_oidit.oid = opiskeluoikeus.oppija_oid
+      LEFT JOIN opiskeluoikeushistoria ON
+        opiskeluoikeushistoria.opiskeluoikeus_id = opiskeluoikeus.id
+        AND opiskeluoikeushistoria.aikaleima < $aikaraja
       WHERE oppija_oid = kaikki_oidit.oid
-         AND koulutusmuoto = ${koulutusmuoto.get}
-         AND suoritustyypit && ${suoritustyypit.get}
-    """.as[KoskiOpiskeluoikeusRow])
-      .flatMap(toResponse(application))
-  }
+        AND NOT mitatoity
```

Comment:
Toimiko vanhakin logiikka valintojen suuntaan muuten mitätöintien osalta täsmälleen samalla tavalla?

Ts. jos opiskeluoikeus on mitätöity, mutta ei ole ollut vielä rajapäivänä mitätöity, niin vanhaankaan versioon ei päästy käsiksi?

Reply:
Mun mielestä valintalaskenta on hakenut aina ekana nykyiset opiskeluoikeusversiot ja kelanut historiaa tarpeeksi taaksepäin, jos opiskeluoikeus on uudempi kuin leikkuripäivä. Pitää varmistaa tämä heiltä, miten haluavat homman toimivan tulevaisuudessa riippumatta siitä miten se nyt toimii.

Reply:
Tämä liittyy myös siihen, mikä on mitätöinnin semantiikka.

Kannattaa ehkä Juhonkin kanssa jutella: voi olla, että tämä nyt toteuttamasi on juuri oikein, koska mitätöinnin oletetaan pyyhkivän aina kaiken historiankin mennessään, eikä sitä historiaa saa olla kellekään Kosken datan jatkohyödyntäjälle enää olemassa.